File: src/repl.rs
//! REPL driver. Binary-only — the language itself (lexer/parser/interp/diagnostic) lives
//! in the `adhoc` library; this module is glue: prompt, history, multi-line continuation,
//! error recovery.
//!
//! History is the one thing kept from the old Common Lisp readline subsystem — `rustyline`
//! gives editing/history/unicode in-process, so the fd-0 invariant, `setlocale` CFFI call,
//! `--has-readline` probe, and `rlwrap` fallback none of them need to exist here.
//!
//! Multi-line continuation: `> ` when nothing is pending, `. ` while continuing. A
//! `ParseError::IncompleteInput` buffers the line for the next round (joined with `\n`);
//! ad's whitespace is insignificant, so nothing else could ever close an unclosed
//! continuation — hence a **blank line cancels** the pending statement rather than being
//! swallowed as whitespace.

use crate::output::{print_eval_error, print_parse_error};
use adhoc::interp::{run, Env};
use adhoc::lexer::{tokenize, TokenKind};
use adhoc::parser::{parse_program, ParseError};
use rustyline::error::ReadlineError;
use rustyline::DefaultEditor;
use std::io::{IsTerminal, Write};
use std::path::PathBuf;

fn history_path() -> PathBuf {
    if let Ok(p) = std::env::var("ADHOC_HISTORY") {
        return PathBuf::from(p);
    }
    let home = std::env::var("HOME").unwrap_or_else(|_| ".".to_string());
    PathBuf::from(home).join(".adhoc_history")
}

/// A blank line, or a line that lexes down to nothing but `--comment` text — i.e. only the
/// sentinel `:eof` token comes out of tokenizing it.
fn is_empty_source(source: &str) -> bool {
    matches!(tokenize(source), Ok(toks) if toks.len() == 1 && toks[0].kind == TokenKind::Eof)
}

enum Input {
    Line(String),
    Eof,
    Interrupted,
}

/// Abstracts over rustyline (interactive TTY) and plain stdin reads (piped input, or
/// rustyline unavailable) behind one read call.
enum Reader {
    Rustyline(Box<DefaultEditor>),
    Stdin,
}

impl Reader {
    fn read(&mut self, prompt: &str) -> Input {
        match self {
            Reader::Rustyline(ed) => match ed.readline(prompt) {
                Ok(line) => {
                    let _ = ed.add_history_entry(line.as_str());
                    Input::Line(line)
                }
                Err(ReadlineError::Eof) => Input::Eof,
                Err(ReadlineError::Interrupted) => Input::Interrupted,
                Err(_) => Input::Eof,
            },
            Reader::Stdin => {
                print!("{prompt}");
                let _ = std::io::stdout().flush();
                let mut buf = String::new();
                match std::io::stdin().read_line(&mut buf) {
                    Ok(0) => Input::Eof,
                    Ok(_) => {
                        if buf.ends_with('\n') {
                            buf.pop();
                            if buf.ends_with('\r') {
                                buf.pop();
                            }
                        }
                        Input::Line(buf)
                    }
                    Err(_) => Input::Eof,
                }
            }
        }
    }
}

pub fn run_repl() {
    let hist_path = history_path();
    // Piped/non-interactive stdin (scripts feeding the REPL, or this binary under test):
    // rustyline still *constructs* successfully but silently drops prompt output rather
    // than erroring, so detecting this by "did rustyline fail to initialize" doesn't work.
    // Check the terminal directly and use the plain-stdin reader whenever it isn't one.
    let mut reader = if std::io::stdin().is_terminal() {
        match DefaultEditor::new() {
            Ok(mut ed) => {
                let _ = ed.load_history(&hist_path);
                Reader::Rustyline(Box::new(ed))
            }
            Err(_) => Reader::Stdin,
        }
    } else {
        Reader::Stdin
    };

    let mut env = Env::new();
    let mut pending = String::new();

    loop {
        let prompt = if pending.is_empty() { "> " } else { ". " };
        match reader.read(prompt) {
            Input::Eof => {
                if !pending.is_empty() {
                    // Re-parse the buffered statement so its (still incomplete) diagnostic
                    // renders instead of exiting silently mid-statement.
                    if let Err(e) = parse_program(&pending) {
                        print_parse_error(&pending, &e);
                    }
                }
                println!();
                break;
            }
            Input::Interrupted => {
                pending.clear();
                continue;
            }
            Input::Line(line) => {
                if !pending.is_empty() && line.trim().is_empty() {
                    pending.clear();
                    println!("-- input cancelled");
                    continue;
                }
                if pending.is_empty() && line.trim().is_empty() {
                    continue;
                }

                let source = if pending.is_empty() { line.clone() } else { format!("{pending}\n{line}") };

                if pending.is_empty() && is_empty_source(&source) {
                    continue;
                }

                match parse_program(&source) {
                    Ok(node) => {
                        pending.clear();
                        match run(&mut env, &node) {
                            Ok(result) => println!("< {}", result.format()),
                            Err(e) => print_eval_error(&source, &e),
                        }
                    }
                    Err(ParseError::IncompleteInput { .. }) => {
                        pending = source;
                    }
                    Err(e) => {
                        pending.clear();
                        print_parse_error(&source, &e);
                    }
                }
            }
        }
    }

    if let Reader::Rustyline(ed) = &mut reader {
        let _ = ed.save_history(&hist_path);
    }
}

File: adhoc/repl.py
"""REPL driver. Binary-side glue over the library: prompt, history, multi-line
continuation, error recovery.

Multi-line continuation: `> ` when nothing is pending, `. ` while continuing. An
`IncompleteInput` parse buffers the line for the next round (joined with `\\n`); ad's
whitespace is insignificant, so nothing else could ever close an unclosed continuation —
hence a **blank line cancels** the pending statement rather than being swallowed as
whitespace.

History lives at `$ADHOC_HISTORY` (default `~/.adhoc_history`) and is only touched on an
interactive terminal; piped stdin (scripts feeding the REPL, tests) goes through plain
reads so the prompt sequence stays observable in transcripts.
"""

import os
import sys

from .compiler import compile_program
from .driver import execute
from .lexer import Eof, LexError, tokenize
from .output import print_eval_error, print_parse_error
from .parser import ALIAS_SEED, IncompleteInput, ParseError, parse_program
from .runtime import EvalError

_EOF = object()
_INTERRUPTED = object()


def _history_path() -> str:
    if p := os.environ.get("ADHOC_HISTORY"):
        return p
    home = os.environ.get("HOME", ".")
    return os.path.join(home, ".adhoc_history")


def _is_empty_source(source: str) -> bool:
    """A blank line, or a line that lexes down to nothing but `--comment` text — i.e.
    only the EOF token comes out of tokenizing it."""
    try:
        toks = tokenize(source)
    except LexError:
        return False
    return len(toks) == 1 and isinstance(toks[0], Eof)


def _read(prompt: str):
    """One input line. input() prints the prompt to stdout and flushes either way, which
    is exactly the observable behavior the Rust Stdin reader had; importing readline
    upgrades it to in-process editing/history when interactive."""
    try:
        return input(prompt)
    except EOFError:
        return _EOF
    except KeyboardInterrupt:
        return _INTERRUPTED


def run_repl(emit_py: bool = False) -> int:
    hist_path = _history_path()
    interactive = sys.stdin.isatty()
    readline = None
    if interactive:
        try:
            import readline as _readline

            readline = _readline
            try:
                readline.read_history_file(hist_path)
            except OSError:
                pass
        except ImportError:
            pass

    env: dict = {}
    consts: set = set()  # user-declared constant names, protected for the session
    modules: dict = {}  # session import registry: `\import` evaluates each file once
    aliases: dict = dict(ALIAS_SEED)  # session alias map; `\alias`/`\dual` extend it
    pending = ""

    while True:
        prompt = ". " if pending else "> "
        got = _read(prompt)

        if got is _EOF:
            if pending:
                # Re-parse the buffered statement so its (still incomplete) diagnostic
                # renders instead of exiting silently mid-statement.
                try:
                    parse_program(pending, aliases, consts)
                except ParseError as e:
                    print_parse_error(pending, e)
            print()
            break

        if got is _INTERRUPTED:
            pending = ""
            continue

        line = got
        if pending and line.strip() == "":
            pending = ""
            print("-- input cancelled")
            continue
        if not pending and line.strip() == "":
            continue

        source = f"{pending}\n{line}" if pending else line

        if not pending and _is_empty_source(source):
            continue

        try:
            node = parse_program(source, aliases, consts)
        except IncompleteInput:
            pending = source
            continue
        except ParseError as e:
            pending = ""
            print_parse_error(source, e)
            continue

        pending = ""
        compiled = compile_program(node)
        if emit_py:
            print(compiled.source, file=sys.stderr)
        try:
            outs = execute(compiled, env, consts, modules)
        except EvalError as e:
            print_eval_error(source, e)
            continue

        if interactive and readline is not None:
            try:
                readline.add_history(line)
            except Exception:
                pass
        # A statement can produce no output (a lone string is a comment-like no-op) —
        # nothing to echo then, not even an empty `< ` line.
        if not outs:
            continue
        print(f"< {outs[-1]}")

    if interactive and readline is not None:
        try:
            readline.write_history_file(hist_path)
        except OSError:
            pass
    return 0
